### PR TITLE
Cache tokenize normalize_function

### PR DESCRIPTION
**Pull request by mrocklin**
This is a modest fraction of overhead when computing many small functions

cc @jcrist

### CODE REVIEW

**Comment by jcrist** [comment]
I'm a bit worried about large cache size here (especially for users using lambdas, which might contain closures, which would then never be cleaned up). What's the context? If it's using the `delayed` api, the solution would be to just decorate the delayed function only once globally, rather than doing it inline multiple times. For dask methods that take in functions, I'd expect the computation and scheduler to dominate the cost of tokenizing a function.

**Comment by mrocklin** [comment]
The context is `futures = [client.submit(inc, i) for i in range(1000)]`

Some options:

1.  We could choose not to cache closures
2.  We could flush the cache from time to time, perhaps every time it reaches length 1000 we remove 500 elements?

**Comment by jreback** [comment]
lru_cache!

**Comment by jcrist** [comment]
> The context is `futures = [client.submit(inc, i) for i in range(1000)]`

Perhaps point to `map` in these contexts?

> We could choose not to cache closures

I don't know of an easy way to check for this (there might be one though), and I'm still wary of unbounded caches. This might be irrational though.

> lru_cache!

Unfortunately `lru_cache` is only in python 3, and we want to support python 2. If we wanted to cache, perhaps use `lru_cache` if available, and fallback to something super simple if not (random evictions as Matt suggested?).

**Comment by mrocklin** [comment]
Related timings

```python
In [1]: def inc(x): return x + 1

In [2]: from dask.base import tokenize

In [3]: %time _ = tokenize(inc)
CPU times: user 459 µs, sys: 29 µs, total: 488 µs
Wall time: 492 µs

In [4]: %timeit tokenize(inc)
The slowest run took 5.21 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 3: 204 µs per loop
```

**Comment by mrocklin** [comment]
Map is good when it's appropriate.  Submit is applicable in more cases.  We can point people to workarounds, yes (delayed is often appropriate), but it's valuable to be a bit idiot proof here.

200us is about how long the scheduler spends on a task in total.

**Comment by mrocklin** [comment]
This also came up in this user-submitted test case on dask.arrays: https://github.com/dask/dask/issues/2034

**Comment by mrocklin** [comment]
I would like for us to consider this again.  @jcrist would any of the solutions proposed above suffice in calming concerns here?

**Comment by mrocklin** [comment]
I've added a very simple cache clearing strategy

File: dask/base.py [hunk 1]
```diff
@@ -269,6 +269,9 @@ def normalize_function(func):
         return function_cache[func]
     except KeyError:
         result = _normalize_function(func)
+        if len(function_cache) > 10000:  # clear half of cache if full
```
Comment (jcrist) [inline dask/base.py]: This seems excessively large. Perhaps a smaller number (say 500?). If the point is to cache for hot loops, I don't think we need to keep around that many.

Comment (mrocklin) [inline dask/base.py]: So reduced.

**Comment by mrocklin** [comment]
Merging.  Thanks for the review @jcrist

**Pull request merged**